File: app/src/lib/sources/romanceio.js
// romance.io discovery source — Books shelf. romance.io is a JS SPA fronted by
// Cloudflare; its private JSON API answers a plain GET from the device (no auth,
// no special header). Like AO3, we call it ON-DEVICE (residential IP) and use
// Supabase only as the store. Books here are METADATA-ONLY: cards link out to
// romance.io, there is nothing to download.
//
// Endpoint: GET /json/topics/books/<include|all>/<sort>/<offset>/<limit>[/<hide>]
//   include / hide = topic slugs, each percent-encoded, joined by literal commas.
//   Slugs are the tree's real _id and often contain SPACES ("grumpy sunshine",
//   "high fantasy") — they must be encodeURIComponent'd (space → %20), NOT hyphenated.
//   offset = (page-1)*limit. The site's page size is 20; the API ignores other limits.
import { fetchJson } from '../fetch.js';
import { ROMANCEIO_DEFAULT_EXCLUDE, ROMANCEIO_TREE } from '../../data/romanceioTopics.js';

// name/slug (lowercased) → canonical slug, for mapping the Books shelf's global
// excluded tags (free-text names in Settings) onto romance.io topics so they can
// be excluded server-side. A name with no matching topic simply can't apply.
const NAME_TO_SLUG = (() => {
  const m = new Map();
  for (const cat of ROMANCEIO_TREE) {
    for (const g of cat.groups) {
      for (const t of g.topics) { m.set(t.s.toLowerCase(), t.s); m.set(t.n.toLowerCase(), t.s); }
    }
  }
  return m;
})();

// Map a list of tag names/slugs to the romance.io slugs that exist (deduped).
export function slugsForNames(names) {
  const out = [];
  for (const n of names || []) {
    const s = NAME_TO_SLUG.get(String(n == null ? '' : (n.name ?? n)).trim().toLowerCase());
    if (s) out.push(s);
  }
  return [...new Set(out)];
}

const HOST = 'https://www.romance.io';
const COVER = 'https://s3.amazonaws.com/romance.io/books'; // /<size>/<_id>.jpg
export const PAGE_SIZE = 20;

// Encode a slug list for a path segment: percent-encode each slug (so spaces and
// any reserved chars are safe), then join with literal commas the API splits on.
const encList = (list) =>
  [...new Set((list || []).map((s) => String(s || '').trim()).filter(Boolean))]
    .map((s) => encodeURIComponent(s))
    .join(',');

// Build the books-endpoint URL. `exclude` is always merged with the permanent
// default-exclude list (non-binary-MC topics the user never wants to see).
export function booksUrl(include, exclude, { sort = 'best', page = 1 } = {}) {
  const inc = encList(include);
  const hide = encList([...(exclude || []), ...ROMANCEIO_DEFAULT_EXCLUDE]);
  const offset = Math.max(0, (Math.max(1, page | 0) - 1) * PAGE_SIZE);
  let path = `/json/topics/books/${inc || 'all'}/${sort}/${offset}/${PAGE_SIZE}`;
  if (hide) path += `/${hide}`;
  return HOST + path;
}

// Normalize one book from the API into the metadata-only match shape the Books
// discovery flow stores. Cover is derived from _id; url links out to romance.io.
function mapBook(b) {
  if (!b || !b._id) return null;
  const id = String(b._id);
  const info = b.info || {};
  const authors = (b.authors || []).map((a) => a && a.name).filter(Boolean);
  const avg = typeof info.avgRating === 'number' ? Math.round(info.avgRating * 10) / 10 : null;
  return {
    sourceWorkId: id,
    title: (info.title || '').trim() || 'Untitled',
    author: authors.join(', ') || 'Unknown',
    summary: (info.description || '').trim(),
    url: b.url ? HOST + b.url : `${HOST}/books/${id}`,
    cover: `${COVER}/large/${id}.jpg`,
    series: (b.series && (b.series.title || b.series.series)) || '',
    rating: avg,
    steam: (info.steam_rating_description || '').trim(),
  };
}

// Fetch one page of books matching the include topics and NOT the exclude topics.
// Returns [] on any failure (best-effort discovery, same as the AO3 path).
export async function searchRomanceBooks(include, exclude, opts = {}) {
  const r = await fetchJson(booksUrl(include, exclude, opts));
  if (!r || r.status < 200 || r.status >= 300 || !r.data || !Array.isArray(r.data.books)) return [];
  return r.data.books.map(mapBook).filter(Boolean);
}

File: worker/ficstash_worker/sources/romanceio.py
"""romance.io discovery source — Books shelf.

romance.io has a private JSON books API that answers a normal server request —
unlike AO3 / Scribble Hub, it does NOT Cloudflare-block datacenter IPs — so its
discovery runs HERE on the worker, like Goodreads. Notify-only: metadata + a link
to romance.io/books/<id>; the user buys the EPUB and uploads it. (The app also
tries romance.io on-device for instant results; both upsert the same tag_matches,
so this is the reliable fallback.)

Endpoint: GET /json/topics/books/<include|all>/<sort>/<offset>/<limit>[/<hide>]
  include / hide = topic SLUGS (each group tag's stored id), percent-encoded and
  comma-joined. Slugs often contain spaces ("grumpy sunshine", "high fantasy").
  A permanent default-exclude list (non-binary-MC topics) is ALWAYS hidden.
"""

from __future__ import annotations

from urllib.parse import quote

import requests

from .base import TAG_SEARCH, WORK_URL, Source, WorkMeta

HOST = "https://www.romance.io"
_UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120 Safari/537.36"
_TIMEOUT = 25
_PAGE = 20

# Non-binary-MC topics the user never wants shown — always appended to hide.
# Mirrors ROMANCEIO_DEFAULT_EXCLUDE in app/src/data/romanceioTopics.js.
DEFAULT_EXCLUDE = [
    "100-plus-years-nbi-mc", "60-100-years-nbi-mc", "afab", "agender", "amab",
    "asexual-nbi-mc", "athlete-nbi-mc", "bodyguard-nbi-mc", "career-professional-nbi-mc",
    "chubby-nbi-mc", "cold-nbi-mc", "cruel-nbi-mc", "demi-sexual-nbi-mc", "early-20s-nbi-mc",
    "famous-nbi-mc", "genderfluid", "immortal-nbi-mc", "in-their-30s", "in-their-40s-50s",
    "late-20s-nbi-mc", "m-f-x", "muscular-nbi-mc", "nerdy-nbi-mc", "non-human-nbi-mc",
    "outlaw-nbi-mc", "poor-nbi-mc", "possessive-nbi-mc", "rake-nbi-mc", "rich-nbi-mc",
    "royal-nbi-mc", "short-nbi-mc", "shy-nbi-mc", "single-parent-nbi-mc", "slim-nbi-mc",
    "stalker-nbi-mc", "sunny-nbi-mc", "sweet-nbi-mc", "tall-nbi-mc", "teacher-nbi-mc",
    "teenage-nbi-mc", "tortured-nbi-mc", "transfemme", "transmasc", "virgin-nbi-mc",
    "warrior-nbi-mc", "working-class-nbi-mc",
]


def _enc(slugs) -> str:
    """Dedupe + percent-encode each slug, join with literal commas the API splits on."""
    seen: list[str] = []
    for s in slugs or []:
        s = (s or "").strip()
        if s and s not in seen:
            seen.append(s)
    return ",".join(quote(s, safe="") for s in seen)


def books_url(include, exclude, page: int = 1) -> str:
    inc = _enc(include) or "all"
    hide = _enc(list(exclude or []) + DEFAULT_EXCLUDE)
    offset = max(0, (max(1, page) - 1) * _PAGE)
    path = f"/json/topics/books/{inc}/best/{offset}/{_PAGE}"
    if hide:
        path += f"/{hide}"
    return HOST + path


def parse_books(data) -> list[WorkMeta]:
    """Pure parser: the books API JSON → [WorkMeta] (metadata-only). Unit-testable."""
    out: list[WorkMeta] = []
    for b in (data or {}).get("books") or []:
        bid = str(b.get("_id") or "")
        if not bid:
            continue
        info = b.get("info") or {}
        authors = [a.get("name") for a in (b.get("authors") or []) if a and a.get("name")]
        tags: list[dict] = []
        avg = info.get("avgRating")
        if isinstance(avg, (int, float)):
            tags.append({"t": f"★ {round(avg, 1)}", "k": "freeform"})
        steam = (info.get("steam_rating_description") or "").strip()
        if steam:
            tags.append({"t": steam, "k": "rating"})
        series = b.get("series") or {}
        out.append(WorkMeta(
            source="romanceio",
            source_work_id=bid,
            title=(info.get("title") or "Untitled").strip() or "Untitled",
            author=", ".join(authors) or "Unknown",
            fandom=(series.get("title") or series.get("series") or ""),
            summary=(info.get("description") or "").strip(),
            tags=tags,
            status="complete",
            url=f"{HOST}/books/{bid}",
        ))
    return out


class RomanceIoSource(Source):
    id = "romanceio"
    # Notify-only, like Goodreads: discovery + a canonical link. No DOWNLOAD.
    capabilities = frozenset({TAG_SEARCH, WORK_URL})

    def __init__(self) -> None:
        self._session = requests.Session()
        self._session.headers["User-Agent"] = _UA

    def work_url(self, source_work_id: str) -> str:
        return f"{HOST}/books/{source_work_id}"

    def search_by_tag(self, tag: str, limit: int = 25) -> list[WorkMeta]:
        return self.search_by_tags([tag], limit=limit)

    def search_by_tags(self, include, exclude=(), limit: int = 25) -> list[WorkMeta]:
        inc = [t for t in (include or []) if t]
        if not inc:
            return []
        try:
            resp = self._session.get(books_url(inc, list(exclude or []), page=1), timeout=_TIMEOUT)
            if resp.status_code != 200:
                return []
            return parse_books(resp.json())[:limit]
        except Exception:  # noqa: BLE001 — a flaky fetch shouldn't kill the sync
            return []
